File: bzb/src/bzb/models/persona.py
"""persona.yaml schema definitions."""
from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=False)


class Identity(_Strict):
    industry: str
    scale: Literal["small", "medium", "large"]
    internal_domain: str
    public_domain: str
    email_pattern: str
    founded: int | None = None
    hq: str | None = None


class NodeRef(_Strict):
    """Top-level entry in persona.yaml's nodes list — references a nodes/<id>.yaml file."""
    id: str
    role: str
    peers: list[str] = Field(default_factory=list)


class Coherence(BaseModel):
    # NOTE: this model uses extra="allow" rather than the strict default
    # so the `world` sub-block introduced 2026-05-18 (Disney rebuild) can
    # carry the full canonical persona-identity tree without each new
    # sub-key requiring a Pydantic class. Once the world schema settles
    # (post Phase 4 of the rebuild — when every lure references it via
    # Jinja and the field set stops churning), replace `extra="allow"`
    # with a typed `World` BaseModel.
    #
    # See: vault/architecture/2026-05-18-crestfield-disney-cohesion-plan.md §1
    model_config = ConfigDict(extra="allow", frozen=False)

    hosts_file_template: str
    ssh_history_seeds: dict[str, list[str]] = Field(default_factory=dict)
    vault_addr: str | None = None
    # IANA timezone the persona "lives in"; mirrors parser.PersonaCoherence.Timezone
    # added in G.1.b. Drives Ollama modified_at, service log timestamps, etc.
    # Empty / unset = UTC.
    timezone: str | None = None


class Persona(_Strict):
    schema_version: Literal[1]
    slug: str
    display_name: str
    version: int = Field(ge=1)
    created_at: date
    last_rotated_at: date
    identity: Identity
    nodes: list[NodeRef] = Field(min_length=1)
    coherence: Coherence
    llm_seed: str | None = None
    lure_content: dict[str, str] = Field(
        default_factory=dict,
        description="persona-specific key-value lure strings (DB creds, service accounts, hostnames, etc.)",
    )

    @model_validator(mode="after")
    def _check_node_graph(self) -> Persona:
        ids = [n.id for n in self.nodes]
        seen: set[str] = set()
        for nid in ids:
            if nid in seen:
                raise ValueError(f"duplicate node id: {nid}")
            seen.add(nid)
        for n in self.nodes:
            for peer in n.peers:
                if peer not in seen:
                    raise ValueError(f"node {n.id} references unknown peer: {peer}")
        return self

File: internal/parser/persona.go
// Package parser — persona.go reads /configurations/persona.yaml at startup.
//
// The persona file is the operator's deception content: company name,
// internal domains, employee email patterns. Beelzebub reads it once and
// passes it to protocol strategies via the builder. Framework code stays
// deception-neutral; persona content lives in the bundle.
package parser

import (
	"fmt"
	"os"
	"path/filepath"

	"gopkg.in/yaml.v3"
)

// Node mirrors the rendered node.yaml that bzb produces under persona/node.yaml
// and mounts into /configurations/node.yaml at runtime.
type Node struct {
	SchemaVersion int         `yaml:"schema_version"`
	NodeID        string      `yaml:"node_id"`
	Role          string      `yaml:"role"`
	PersonaLocal  NodePersona `yaml:"persona_local"`
	Lures         []string    `yaml:"lures"`
	CanarySlots   []string    `yaml:"canary_slots,omitempty"`
}

// NodePersona holds the runtime-visible server identity and resource data.
// Node-prefixed to avoid collision with shellemulator's own Process/Listener/NetworkConfig types.
type NodePersona struct {
	Hostname    string              `yaml:"hostname"`
	FQDN        string              `yaml:"fqdn"`
	OS          string              `yaml:"os"`
	User        string              `yaml:"user"`
	InternalIP  string              `yaml:"internal_ip"`
	UptimeDays  int                 `yaml:"uptime_days"`
	ProcessSeed string              `yaml:"process_seed"`
	Processes   []NodeProcess       `yaml:"processes,omitempty"`
	EnvVars     map[string]string   `yaml:"env_vars,omitempty"`
	Lures       map[string]string   `yaml:"lures,omitempty"`
	Filesystem  map[string][]string `yaml:"filesystem,omitempty"`
	Network     *NodeNetworkConfig  `yaml:"network,omitempty"`
	Listeners   []NodeListener      `yaml:"listeners,omitempty"`
}

// NodeProcess represents a single row in the process table (ps aux output).
type NodeProcess struct {
	PID  int    `yaml:"pid"`
	User string `yaml:"user"`
	CPU  string `yaml:"cpu"`
	Mem  string `yaml:"mem"`
	VSZ  string `yaml:"vsz"`
	RSS  string `yaml:"rss"`
	Cmd  string `yaml:"cmd"`
	Stat string `yaml:"stat"`
	Time string `yaml:"time"`
}

// NodeListener represents a listening socket entry (netstat/ss output).
type NodeListener struct {
	Proto   string `yaml:"proto"`
	Local   string `yaml:"local"`
	PID     int    `yaml:"pid"`
	Program string `yaml:"program"`
}

// NodeNetworkConfig holds network interface and routing configuration.
type NodeNetworkConfig struct {
	Interface string `yaml:"interface"`
	MAC       string `yaml:"mac"`
	IP        string `yaml:"ip"`
	Netmask   string `yaml:"netmask"`
	Broadcast string `yaml:"broadcast"`
	Gateway   string `yaml:"gateway"`
}

// LoadNode reads <dir>/node.yaml — the rendered single-node descriptor
// that bzb produces under persona/node.yaml and mounts at /configurations/node.yaml.
func LoadNode(dir string) (*Node, error) {
	data, err := os.ReadFile(filepath.Join(dir, "node.yaml"))
	if err != nil {
		return nil, fmt.Errorf("read node.yaml: %w", err)
	}
	var n Node
	if err := yaml.Unmarshal(data, &n); err != nil {
		return nil, fmt.Errorf("parse node.yaml: %w", err)
	}
	return &n, nil
}

// Persona mirrors the bzb-side schema. Carries only fields beelzebub uses
// at runtime; extra YAML fields are intentionally ignored so future schema
// additions don't break older binaries.
type Persona struct {
	SchemaVersion int              `yaml:"schema_version"`
	Slug          string           `yaml:"slug"`
	DisplayName   string           `yaml:"display_name"`
	Version       int              `yaml:"version"`
	Identity      PersonaIdentity  `yaml:"identity"`
	Coherence     PersonaCoherence `yaml:"coherence"`
	LLMSeed       string           `yaml:"llm_seed,omitempty"`
	// LureContent holds persona-specific key-value deception strings
	// (DB credentials, service accounts, internal hostnames, etc.)
	// so Go code stays deception-neutral and references values by key.
	LureContent map[string]string `yaml:"lure_content,omitempty"`
}

// Lure returns the value for the given key from LureContent, or "" if not set.
// Safe to call on a nil Persona.
func (p *Persona) Lure(key string) string {
	if p == nil || p.LureContent == nil {
		return ""
	}
	return p.LureContent[key]
}

// PersonaIdentity holds the canonical identity fields used by protocol handlers.
type PersonaIdentity struct {
	Industry       string `yaml:"industry"`
	Scale          string `yaml:"scale"`
	InternalDomain string `yaml:"internal_domain"`
	PublicDomain   string `yaml:"public_domain"`
	EmailPattern   string `yaml:"email_pattern"`
	Founded        int    `yaml:"founded,omitempty"`
	HQ             string `yaml:"hq,omitempty"`
}

// PersonaCoherence holds coherence/consistency data used across protocols.
type PersonaCoherence struct {
	HostsFileTemplate string              `yaml:"hosts_file_template"`
	SSHHistorySeeds   map[string][]string `yaml:"ssh_history_seeds,omitempty"`
	VaultAddr         string              `yaml:"vault_addr,omitempty"`
	// Timezone is the IANA timezone name (e.g. "Asia/Singapore", "UTC")
	// the persona "lives in". Drives wall-clock formatting on every emitted
	// timestamp that should be persona-coherent (Ollama modified_at,
	// service log timestamps, etc.). Empty = UTC.
	Timezone string `yaml:"timezone,omitempty"`
}

// LoadPersona reads <dir>/persona.yaml and returns the parsed struct.
// dir is typically "/configurations" (mounted into beelzebub at runtime).
func LoadPersona(dir string) (*Persona, error) {
	path := filepath.Join(dir, "persona.yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read persona.yaml: %w", err)
	}
	var p Persona
	if err := yaml.Unmarshal(data, &p); err != nil {
		return nil, fmt.Errorf("parse persona.yaml: %w", err)
	}
	if p.SchemaVersion != 1 {
		return nil, fmt.Errorf("unsupported schema_version: %d", p.SchemaVersion)
	}
	if p.Slug == "" {
		return nil, fmt.Errorf("persona.yaml missing required field: slug")
	}
	return &p, nil
}
